lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,546 @@
+//! payment a fixed number of times before failing or succeeding as needed.
+//!
+//! # Example
+//!
+//! TODO: Fill this in once the API is finalized.
+//!
+//! # Note
+//!
+//! The [`Route`] is compute before each payment attempt. Any updates affecting path finding such as
+//! updates to the network graph or changes to channels scores should be applied prior to retries.
+//! This typically means any [`EventHandler`] decorators responsible for this should decorate
+//! [`InvoicePayer`] so that such changes take effect before retrying.
+
+use crate::Invoice;
+
+use bitcoin::hashes::Hash;
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::secp256k1::key::PublicKey;
+
+use lightning::ln::{PaymentHash, PaymentSecret};
+use lightning::ln::channelmanager::{ChannelDetails, PaymentSendFailure};
+use lightning::ln::features::InvoiceFeatures;
+use lightning::ln::msgs::LightningError;
+use lightning::routing::router::{Route, RouteHint};
+use lightning::util::events::{Event, EventHandler};
+use lightning::util::logger::Logger;
+
+use std::collections::hash_map::{self, HashMap};
+use std::ops::Deref;
+use std::sync::Mutex;
+
+/// A utility for paying [`Invoice]`s.
+pub struct InvoicePayer<P: Deref, R, L: Deref, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	payer: P,
+	router: R,
+	logger: L,
+	event_handler: E,
+	invoice_cache: Mutex<HashMap<PaymentHash, (Invoice, usize)>>,
+	retry_attempts: usize,
+}
+
+/// A trait defining behavior of an [`Invoice`] payer.
+pub trait Payer {
+	/// Returns the payer's node id.
+	fn node_id(&self) -> PublicKey;
+
+	/// Returns the payer's channels.
+	fn first_hops(&self) -> Vec<ChannelDetails>;
+
+	/// Sends a payment over the Lightning Network using the given [`Route`].
+	fn send_payment(
+		&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>
+	) -> Result<(), PaymentSendFailure>;
+}
+
+/// A trait defining behavior for routing an [`Invoice`] payment.
+pub trait Router {
+	/// Finds a [`Route`] between `payer` and `payee` using the provided first and last hops for a
+	/// payment with the given values.
+	fn find_route(
+		&self,
+		payer: &PublicKey,
+		payee: &PublicKey,
+		payee_features: Option<InvoiceFeatures>,
+		first_hops: Option<&[&ChannelDetails]>,
+		last_hops: &[&RouteHint],
+		final_value_msat: u64,
+		final_cltv: u32,
+	) -> Result<Route, LightningError>;
+}
+
+/// An error that may occur when making a payment.
+#[derive(Debug)]
+pub enum PaymentError {
+	/// An error resulting from the provided [`Invoice`] or payment hash.
+	Invoice(&'static str),
+	/// An error occurring when finding a route.
+	Routing(LightningError),
+	/// An error occurring when sending a payment.
+	Sending(PaymentSendFailure),
+}
+
+impl<P: Deref, R, L: Deref, E> InvoicePayer<P, R, L, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	/// Creates an invoice payer.
```

Comment:
Might be missing it, but would be good to document the default number of retries